### PR TITLE
WIP Neo

### PR DESCRIPTION
**Pull request by rodoufu**
## Description

Implemented the NEO transaction and signing.

## Testing instructions

`make test --gtest_filter=NEO*`

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [x] Add the coin definition to coins.json and TWCoinType.
- [x] Implement functionality in C++. Put it in a subfolder of src/NEO.
  - [x] Address (if necessary)
  - [x] Transaction (if necessary)
  - [x] Signer
- [x] Write unit tests. Put them in a subfolder of tests/NEO.
  - [ ] Mnemonic phrase - > Address derivation test. Put this test in the CoinTests.cpp file
  - [ ] Transaction signing tests
- [ ] Add relevant constants in TWP2SHPrefix, TWEthereymChainID, TWHRP, etc., as necessary.
- [x] Implement address validation and derivation
  - [x] Implement validation and derivation in src/Coin.cpp.
  - [x] Add tests for validation in tests/CoinAddressValidationTests.cpp and derivation in tests/CoinAddressDerivationTests.cpp. 
- [ ] Write interface header in include/TrustWalletCore and implement the interface in src/interface.
  - [x] Address interface (if necessary).
  - [ ] Signing interface.
- [ ] Validate generated code in Android an iOS projects. Write integration tests for each.
- [ ] Add a coin icon to section in docs/coins.md
  - [ ] Upload coin icon to TrustWallet/tokens if necessary

### CODE REVIEW

**Comment by hewigovens** [comment]
Hi, could you please create a clean pr? too many unrelated changes, it's not easy review

**Comment by rodoufu** [comment]
> Hi, could you please create a clean pr? too many unrelated changes, it's not easy review

Yeah, I know, sorry about that.
But I'm facing some trouble with my environment, I'm not being able to `tools/ios-test` here.
What do you suggest me to do?
I thought about `git rebase -i ` but I'm a little afraid about that.

File: src/NEO/Address.h [hunk 1]
```diff
@@ -13,25 +13,31 @@
 #include <string>
 
 namespace TW::NEO {
+    static const unsigned AddressSize{21};
 
-class Address : public TW::Base58Address<21> {
```
Comment (hewigovens) [inline src/NEO/Address.h]: can you follow existing code styles? no indentation for class definition, and two spaces for `public`

File: src/NEO/CoinReference.hpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#include <vector>
+
+#include "ISerializable.h"
+#include "../uint256.h"
+#include "../UInt.hpp"
+#include "../Data.h"
+#include "../ReadData.h"
+
+namespace TW::NEO {
+    class CoinReference : public ISerializable {
```
Comment (hewigovens) [inline src/NEO/CoinReference.hpp]: same here, can you remove the indentation?, please also add license header to each file

File: src/NEO/OpCode.h [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+#include <cstdint>
+
+/// https://github.com/neo-project/neo-vm/blob/master/src/neo-vm/OpCode.cs
+namespace TW::NEO {
+
+/// An empty array of bytes is pushed onto the stack.
+    static const uint8_t PUSH0{0x00};
+    static const uint8_t PUSHF{PUSH0};
+//static const uint8_t PUSH_BYTE_33{0x21};
```
Comment (hewigovens) [inline src/NEO/OpCode.h]: remove not used code

File: src/NEO/Signer.cpp [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    return Address(address);
+}
+
+Data Signer::sign(Transaction &tx) const {
+    return sign(tx.getHash());
+}
+
+Data Signer::sign(const Data &data) const {
+    auto signature = getPrivateKey().sign(TW::Hash::sha256(data), TWCurveNIST256p1);
+    signature.pop_back();
```
Comment (hewigovens) [inline src/NEO/Signer.cpp]: why we call pop_back here?

File: src/NEO/Signer.h [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace TW::NEO {
+    class Signer {
+    private:
+        Data publicKey;
+        TW::PrivateKey privateKey;
+        std::string address;
+    public:
+        explicit Signer(const TW::PrivateKey &priKey);
+        PrivateKey getPrivateKey() const;
```
Comment (hewigovens) [inline src/NEO/Signer.h]: `getPrivateKey`, `getPublicKey` and `getAddress` are quite trivial, can we remove it from signer?

File: src/NEO/Transaction.cpp [hunk 6]
```diff
@@ -0,0 +1,97 @@
+#include <ctype.h>
```
Comment (hewigovens) [inline src/NEO/Transaction.cpp]: Add license header please

File: src/NEO/Transaction.cpp [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+#include "../uint256.h"
+#include "../UInt.hpp"
```
Comment (hewigovens) [inline src/NEO/Transaction.cpp]: for consistency, can we keep using `.h`? not `.hpp`

File: src/NEO/Transaction.cpp [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    for (auto &__x_ : (X)) { \
+        (V) += __x_.size(); \
+    } \
+}
+int64_t Transaction::size() const {
+    auto resp = 1 + 1; // sizeof(type) + sizeof(version);
+    SIZE_INC(resp, attributes);
+    SIZE_INC(resp, inInputs);
+    SIZE_INC(resp, outputs);
+//    SIZE_INC(resp, witnesses); // witnessess is not serialized
```
Comment (hewigovens) [inline src/NEO/Transaction.cpp]: should we add a flag to serialize witness data too?

File: src/NEO/Transaction.cpp [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    initial_pos = deserializeExclusiveData(data, initial_pos);
+    attributes.clear();
+    initial_pos = ISerializable::deserialize<TransactionAttribute>(attributes, data, initial_pos);
+    inInputs.clear();
+    initial_pos = ISerializable::deserialize<CoinReference>(inInputs, data, initial_pos);
+    outputs.clear();
+    ISerializable::deserialize<TransactionOutput>(outputs, data, initial_pos);
+
+    // witnesses.clear();
+    // initial_pos = ISerializable::deserialize<Witness>(witnesses, data, initial_pos);
```
Comment (hewigovens) [inline src/NEO/Transaction.cpp]: or just remove not used code

File: src/NEO/TransactionAttribute.hpp [hunk 10]
```diff
@@ -0,0 +1,53 @@
+#pragma once
```
Comment (hewigovens) [inline src/NEO/TransactionAttribute.hpp]: license header and `.h`

File: src/ReadData.cpp [hunk 11]
```diff
@@ -0,0 +1,133 @@
+#include "Data.h"
+#include "ReadData.h"
```
Comment (hewigovens) [inline src/ReadData.cpp]: can you reuse the code from `BinaryCoding.h`?

**Comment by hewigovens** [comment]
you don't implement c interface for neo signer, please also add signer tests in js/kotlin/swift

File: src/NEO/CoinReference.hpp (hunk 2 above)
Comment (rodoufu) [reply src/NEO/CoinReference.hpp]: Sure, I'm going to add the license to all files.

File: src/NEO/Address.h (hunk 1 above)
Comment (rodoufu) [reply src/NEO/Address.h]: I'm removing the indentation for all classes, when you say two spaces for `public`, do you mean for all kinds of indentation?

File: src/NEO/Signer.cpp (hunk 4 above)
Comment (rodoufu) [reply src/NEO/Signer.cpp]: The signing generates an unused byte (zero), it's the same behavior on the Ontology as you may see in:
https://github.com/TrustWallet/wallet-core/blob/master/src/Ontology/Signer.cpp#L50
In a matter of fact, Ontology looks a lot with NEO.

File: src/NEO/Transaction.cpp (hunk 6 above)
Comment (rodoufu) [reply src/NEO/Transaction.cpp]: Done.

File: src/NEO/Transaction.cpp (hunk 7 above)
Comment (rodoufu) [reply src/NEO/Transaction.cpp]: Done for all files.

File: src/ReadData.cpp (hunk 11 above)
Comment (rodoufu) [reply src/ReadData.cpp]: I'm going to take a look at it.

**Comment by rodoufu** [comment]
@hewigovens Just like Ethereum ERC20 tokens, Neo has its NEP5 tokens, how Trust Wallet deals with that?
Do we need to register each token in order for it to be available? Or just implementing Neo is enough?

**Comment by hewigovens** [comment]
hey @rodoufu , are you still working on this?

**Comment by rodoufu** [comment]
> hey @rodoufu , are you still working on this?

Hey @hewigovens I'm.
I'm a little out of time these days but I'll continue it next week.

**Comment by rodoufu** [comment]
> you don't implement c interface for neo signer, please also add signer tests in js/kotlin/swift

@hewigovens What do you mean by that? Could you give me some help with that? I've created the following code on the Signer.h, what else do I need to do?
```c
/// Wrapper for C interface.
struct TWNEOSigner {
    TW::NEO::Signer impl;
};

```

**Comment by hewigovens** [comment]
close for a new clean pr now

**Comment by vncoelho** [comment]
Hi Trust Wallet team, how are you?

     Rodoufu has been working kind of alone on this.

     In general, how are these inclusions dealt?
     Can someone from TrustWallet team help with the UT's for IOS and Android? The replication for js/kotlin/swift? Or, in theory, it should all come done in the PR?

**Comment by vikmeup** [comment]
Nebulas: https://github.com/TrustWallet/wallet-core/pull/519
Waves: https://github.com/TrustWallet/wallet-core/pull/489

**Comment by vikmeup** [comment]
@vncoelho please finish re-open when integration is complete.

For the reference here is an example of the completed PR: 

Nebulas: https://github.com/TrustWallet/wallet-core/pull/519
Waves: https://github.com/TrustWallet/wallet-core/pull/489
Aeternity: https://github.com/TrustWallet/wallet-core/pull/420

Full integration guide is available here: https://developer.trustwallet.com/wallet-core/newblockchain

**Pull request closed without merging**